js/layout-javascript/agenda-code.js

```diff
@@ -1239,6 +1245,7 @@ DynamicList.prototype.renderLoopHTML = function () {
         _this.$container.find('.agenda-list-day-holder').eq(0).addClass('active');
       }
 
+      // Break render cycle if there is no more data
       if (!nextBatch.length) {
         resolve(_.flatten(_this.getAgendasByDay()));
```

Comment:
I'm recommending the previous `if` be refactored to remove the `-1` case. If this is done, let's move this `if` inside that `if` block.

Also, if this is the first iteration and there's no data at all, does this mean we can just use the below for simplicity?

```suggestion
        resolve([]);
```

Reply:
@tonytlwu 
Could you please explain the first sentence, I did not understand. Do you want to put the if statement from line 1249 inside if statement line 1243?

Reply:
@MaksymShokin Let me know if this works.

```js
  _this.$container.find('.new-agenda-list-container').removeClass('loading').addClass('ready');
  _this.$container.find('.agenda-list-day-holder').eq(0).addClass('active');

  return new Promise(function (resolve) {
    function render() {
      // Get the next batch of items to render
      var nextBatch = _this.getAgendasByDay().slice(
        _this.INCREMENTAL_RENDERING_BATCH_SIZE * renderLoopIndex,
        _this.INCREMENTAL_RENDERING_BATCH_SIZE * (renderLoopIndex + 1)
      );

      // Break render cycle if there is no more data
      if (!nextBatch.length && renderLoopIndex > 0) {
        resolve(_.flatten(_this.getAgendasByDay()));
        return;
      }

      var $renderBatch = $(template(nextBatch));

      $renderFull.add($renderBatch);
      $agendaListHolder.append($renderBatch);

      // Break render cycle if there is no data at all
      if (!nextBatch.length) {
        resolve([]);
        return;
      }

      renderLoopIndex++;

      // if the browser is ready, render
      requestAnimationFrame(render);
    }
    // start the initial render
    requestAnimationFrame(render);
  });
```

Reply:
@tonytlwu 
Seems to work correctly

Reply:
@MaksymShokin Can you see if this also works? I realised looking at my code snippet that the second cycle break isn't necessary since I've moved the initial UI state switching outside of the render loop.

```js
  _this.$container.find('.new-agenda-list-container').removeClass('loading').addClass('ready');
  _this.$container.find('.agenda-list-day-holder').eq(0).addClass('active');

  return new Promise(function (resolve) {
    function render() {
      // Get the next batch of items to render
      var nextBatch = _this.getAgendasByDay().slice(
        _this.INCREMENTAL_RENDERING_BATCH_SIZE * renderLoopIndex,
        _this.INCREMENTAL_RENDERING_BATCH_SIZE * (renderLoopIndex + 1)
      );

      // Break render cycle if there is no more data
      if (!nextBatch.length) {
        resolve(_.flatten(_this.getAgendasByDay()));
        return;
      }

      var $renderBatch = $(template(nextBatch));

      $renderFull.add($renderBatch);
      $agendaListHolder.append($renderBatch);

      renderLoopIndex++;

      // if the browser is ready, render
      requestAnimationFrame(render);
    }
    // start the initial render
    requestAnimationFrame(render);
  });
```

Reply:
fixed